src/main/java/org/opensearch/security/privileges/PrivilegesEvaluatorResponse.java

```diff
@@ -26,33 +26,115 @@
 
 package org.opensearch.security.privileges;
 
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
 import java.util.HashSet;
+import java.util.List;
 import java.util.Set;
 
+import com.google.common.collect.ImmutableSet;
+
 import org.opensearch.action.admin.indices.create.CreateIndexRequestBuilder;
 
+import com.selectivem.collections.CheckTable;
+
 public class PrivilegesEvaluatorResponse {
     boolean allowed = false;
-    Set<String> missingPrivileges = new HashSet<String>();
     Set<String> missingSecurityRoles = new HashSet<>();
-    Set<String> resolvedSecurityRoles = new HashSet<>();
     PrivilegesEvaluatorResponseState state = PrivilegesEvaluatorResponseState.PENDING;
     CreateIndexRequestBuilder createIndexRequestBuilder;
-
+    private Set<String> onlyAllowedForIndices = ImmutableSet.of();
+    private CheckTable<String, String> indexToActionCheckTable;
+    private String privilegeMatrix;
+    private String reason;
+
+    /**
+     * Contains issues that were encountered during privilege evaluation. Can be used for logging.
+     */
+    private List<PrivilegesEvaluationException> evaluationExceptions = new ArrayList<>();
+
+    /**
+     * Returns true if the request can be fully allowed. See also isAllowedForSpecificIndices().
+     */
     public boolean isAllowed() {
         return allowed;
     }
 
+    /**
+     * Returns true if the request can be allowed if the referenced indices are reduced (aka "do not fail on forbidden").
+     * See getAvailableIndices() for the indices for which we have privileges.
+     */
+    public boolean isPartiallyOk() {
+        return !this.onlyAllowedForIndices.isEmpty();
+    }
+
+    /**
+     * In case isPartiallyOk() is true, this returns the indices for which we have privileges.
+     */
+    public Set<String> getAvailableIndices() {
+        return this.onlyAllowedForIndices;
+    }
+
+    /**
+     * In case isAllowed() is false, this returns the privileges (aka action names) for which we do not have sufficient
+     * privileges.
+     */
     public Set<String> getMissingPrivileges() {
-        return new HashSet<String>(missingPrivileges);
+        return this.indexToActionCheckTable != null ? this.indexToActionCheckTable.getIncompleteColumns() : Collections.emptySet();
```

Comment:
This will always be empty if the checkTable is null? In what scenarios can the checktable be null?

Reply:
I think this is the last remaining case where `checktable` is null:

https://github.com/opensearch-project/security/blob/1fef11be731e5e2a3c06c5d565faf23525ffe0d6/src/main/java/org/opensearch/security/privileges/PrivilegesEvaluator.java#L328-L343

Indeed, the concept of missing privileges in not applicable there.

Of course, I can also try to refactor this, but I am not sure if that's really in the scope of the PR.

Reply:
Should we be addressing this in a follow-up? Seems like a good piece to refactor

Reply:
Yes, cleaning up PrivilegeEvaluatorResponse is a good idea!